File: app/src/main/java/org/wen/gank/CommonTabAdapter.java
package org.wen.gank;

import android.content.Context;
import android.support.annotation.NonNull;
import android.support.annotation.StringRes;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentPagerAdapter;
import android.support.v4.view.ViewPager;
import java.util.ArrayList;
import java.util.List;

/**
 * created by Jiahui.wen 2017-07-23
 */
public class CommonTabAdapter extends FragmentPagerAdapter {

    private List<FragmentPagerItem> pagerItems;

    public CommonTabAdapter(FragmentManager fm) {
        super(fm);
    }

    @Override
    public Fragment getItem(int position) {
        return pagerItems.get(position).fragment;
    }

    @Override
    public int getCount() {
        return pagerItems != null ? pagerItems.size() : 0;
    }

    @Override
    public CharSequence getPageTitle(int position) {
        return pagerItems.get(position).title;
    }

    void setData(List<FragmentPagerItem> pagerItems) {
        this.pagerItems = pagerItems;
    }

    //=============================================================
    // Builder
    //=============================================================

    public static class Builder {

        private final Context context;
        private final FragmentManager fragmentManager;

        private List<FragmentPagerItem> fragmentItems = new ArrayList<>();

        public Builder(Context context, FragmentManager fragmentManager) {
            this.context = context;
            this.fragmentManager = fragmentManager;
        }

        public Builder add(@StringRes int title, @NonNull Fragment fragment) {
            return add(context.getString(title), fragment);
        }

        public Builder add(String title, @NonNull Fragment fragment) {
            FragmentPagerItem item = new FragmentPagerItem(title, fragment);
            fragmentItems.add(item);
            return this;
        }

        public CommonTabAdapter into(@NonNull ViewPager viewPager) {
            CommonTabAdapter adapter = new CommonTabAdapter(fragmentManager);
            adapter.setData(fragmentItems);
            viewPager.setAdapter(adapter);
            return adapter;
        }
    }

    private static class FragmentPagerItem {

        public final String title;
        public final Fragment fragment;

        public FragmentPagerItem(String title, Fragment fragment) {
            this.title = title;
            this.fragment = fragment;
        }
    }
}

File: app/src/main/java/org/wen/gank/CommonTabAdapter.kt
package org.wen.gank

import android.content.Context
import androidx.annotation.StringRes
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentManager
import androidx.fragment.app.FragmentPagerAdapter
import androidx.viewpager.widget.ViewPager
import java.util.*

/**
 * created by Jiahui.wen 2017-07-23
 */
class CommonTabAdapter(fm: FragmentManager) : FragmentPagerAdapter(fm) {

  private var pagerItems: List<FragmentPagerItem>? = null

  override fun getItem(position: Int): Fragment {
    return pagerItems!![position].fragment
  }

  override fun getCount(): Int {
    return if (pagerItems != null) pagerItems!!.size else 0
  }

  override fun getPageTitle(position: Int): CharSequence? {
    return pagerItems!![position].title
  }

  fun setData(pagerItems: List<FragmentPagerItem>) {
    this.pagerItems = pagerItems
  }

  //=============================================================
  // Builder
  //=============================================================

  class Builder(private val context: Context, private val fragmentManager: FragmentManager) {

    private val fragmentItems = ArrayList<FragmentPagerItem>()

    fun add(@StringRes title: Int, fragment: Fragment): Builder {
      return add(context.getString(title), fragment)
    }

    fun add(title: String, fragment: Fragment): Builder {
      val item = FragmentPagerItem(title, fragment)
      fragmentItems.add(item)
      return this
    }

    fun into(viewPager: ViewPager): CommonTabAdapter {
      val adapter = CommonTabAdapter(fragmentManager)
      adapter.setData(fragmentItems)
      viewPager.adapter = adapter
      return adapter
    }
  }

  class FragmentPagerItem(val title: String, val fragment: Fragment)
}
